lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -16,30 +17,48 @@ module Cardano.Wallet.Primitive.AddressDiscovery.Random
     (
     -- ** State
       RndState (..)
-    , mkRndState
     ) where
 
 import Prelude
 
+import Cardano.Byron.Codec.Cbor
+    ( decodeAddressDerivationPath, decodeAddressPayload, deserialise )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..), XPrv )
 import Cardano.Wallet.Primitive.AddressDerivation.Random
-    ( RndKey )
+    ( RndKey (..) )
 import Cardano.Wallet.Primitive.AddressDiscovery
     ( CompareDiscovery (..)
     , GenChange (..)
     , IsOurs (..)
     , IsOwned (..)
     , KnownAddresses (..)
     )
-import Data.ByteString
-    ( ByteString )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeDeserialiseCbor, unsafeFromHex )
+import Control.DeepSeq
+    ( NFData )
+import GHC.Generics
+    ( Generic )
 
-newtype RndState = RndState ByteString
+import qualified Data.ByteString.Lazy as BL
 
-mkRndState :: RndState
-mkRndState = error "AddressDiscovery.Random unimplemented"
+
+newtype RndState = RndState { getRndState :: RndKey 'RootK XPrv }
+    deriving (Generic)
+
+instance NFData RndState
 
 instance IsOurs RndState where
-    isOurs _ s = (False, s)
+    isOurs (Address addr) (RndState key) = do
+        let payload = unsafeDeserialiseCbor decodeAddressPayload $
```

Comment:
Why `unsafe` here? It's totally possible that we would call `IsOurs` on addresses that aren't actually in a format we expect. This should rather be gracefully handled (i.e., if we can't even parse the address, it's most likely not ours :grimacing: )

Reply:
Fixed.